walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -23,41 +23,74 @@
+    if (meetingDuration > TimeRange.WHOLE_DAY.duration()) {
       return Collections.emptyList();
     }
     if (events.isEmpty()) {
       return Arrays.asList(TimeRange.WHOLE_DAY);
     }
+    Collection<String> mandatoryMeetingAttendees = request.getAttendees();
+    Collection<String> optionalMeetingAttendees = request.getOptionalAttendees();
     Collection<TimeRange> meetingTimes = new TreeSet<>(TimeRange.ORDER_BY_START);
     meetingTimes.add(TimeRange.WHOLE_DAY);
+    Collection<Event> optionalEvents = new ArrayList<>();
     for (Event event : events) {
       Collection<String> eventAttendees = event.getAttendees();
-      Collection<String> meetingAttendees = request.getAttendees();
-      if (Collections.disjoint(eventAttendees, meetingAttendees)) {
+      if (Collections.disjoint(eventAttendees, mandatoryMeetingAttendees)) {
+        if (!Collections.disjoint(eventAttendees, optionalMeetingAttendees)) {
+          optionalEvents.add(event);
+        }
         continue;
       }
-      TimeRange eventTime = event.getWhen();
-      for (TimeRange meetingTime : meetingTimes) {
-        if (eventTime.overlaps(meetingTime)) {
-          meetingTimes.remove(meetingTime);
-          if (meetingTime.start() < eventTime.start()) {
-            TimeRange before = TimeRange.fromStartEnd(meetingTime.start(), eventTime.start(), false);
-            if (before.duration() >= request.getDuration()) {
-              meetingTimes.add(before);
-            }
-          }
-          if (eventTime.end() < meetingTime.end()) {
-            TimeRange after = TimeRange.fromStartEnd(eventTime.end(), meetingTime.end(), false);
-            if (after.duration() >= request.getDuration()) {
-              meetingTimes.add(after);
-            }
-          }
-        }
+      meetingTimes = splitMeetingTimes(event, meetingTimes, meetingDuration);
+      if (meetingTimes.isEmpty()) {
+        break;
       }
     }
+
+    Collection<TimeRange> optionalMeetingTimes = new TreeSet<>(TimeRange.ORDER_BY_START);
+    optionalMeetingTimes.addAll(meetingTimes);
+    for (Event optionalEvent : optionalEvents) {
+      optionalMeetingTimes = splitMeetingTimes(optionalEvent, optionalMeetingTimes, meetingDuration);
+      if (optionalMeetingTimes.isEmpty()) {
+        break;
+      }
+    }
+
+    if (mandatoryMeetingAttendees.isEmpty()) {
+      meetingTimes = optionalMeetingTimes;
+    } else {
+      meetingTimes = optionalMeetingTimes.isEmpty() ? meetingTimes : optionalMeetingTimes;
+    }
+
     Collection<TimeRange> finalMeetingTimes = new ArrayList<>();
     finalMeetingTimes.addAll(meetingTimes);
     return finalMeetingTimes;
   }
+
+  public Collection<TimeRange> splitMeetingTimes(Event event, Collection<TimeRange> meetingTimes, long meetingDuration) {
+    Collection<TimeRange> newMeetingTimes = new TreeSet<>(TimeRange.ORDER_BY_START);
+    TimeRange eventTime = event.getWhen();
+    for (TimeRange meetingTime : meetingTimes) {
```

Comment:
Maybe I'm missing something but it seems like each computation on each element is independent so can't we also make the inside of the for loop it's own function?  This should help with the nesting.